TRACKER-225 Make DiscoveryMetadataClient shared across threads of a handler.

anwar6953 opened the pull request:
Make DiscoveryMetadataClient shared across threads of a handler, so that each handler doesn't open too many zookeeper connections.

https://issues.cask.co/browse/TRACKER-225

poornachandra commented on src/main/java/co/cask/tracker/AuditTagsHandler.java:
```diff
@@ -80,13 +81,17 @@
+
+  // This is static to be shared across threads of the handler.
+  // We don't want to have each thread having its own zkClient (connection). See TRACKER-225.
+  private static final AtomicReference<DiscoveryMetadataClient> discoveryMetadataClientRef = new AtomicReference<>();
```
Is the `DiscoveryMetadataClient` thread-safe?

poornachandra replied:
Would it be simpler to have `DiscoveryMetadataClient` take in a `ZookeeperClient` instance in its constructor?

anwar6953 replied:
Seems like `DiscoveryMetadataClient` is thread-safe, because the only fields it as are:
`clientConfig`, which is read-only
`endpointStrategySupplier`, wrapper around a ZKDiscoveryClient, which is also thread-safe.

poornachandra replied:
Looks like `ClientConfig` is not thread safe since it has mutable state that can change without synchronization. One simple option is to change `ClientConfig` to be immutable so that it becomes thread-safe.

anwar6953 replied:
https://github.com/caskdata/cdap/pull/6751

chtyim commented on src/main/java/co/cask/tracker/AuditTagsHandler.java:
```diff
@@ -267,10 +271,16 @@ private ZKClientService createZKClient(String zookeeperQuorum) {
+    DiscoveryMetadataClient discoveryMetadataClient = DISCOVERY_METADATA_CLIENT_REF.get();
     if (discoveryMetadataClient == null) {
-      this.discoveryMetadataClient = createMetadataClient(request);
+      synchronized (DISCOVERY_METADATA_CLIENT_REF) {
```
If you use synchronized, no need to use atomic reference. You can use volatile + synchronize to avoid synchronizing on every get call as well.

```java
static volatile DiscoveryMetadataClient client;
...
DiscoveryMetadataClient c = client;
if (c != null) {
  return c;
}
synchronize (AuditTagsHandler.class) {
  c = client;
  if (c != null) {
    return c;
  }
  client = c = createDiscoveryMetaClient();
}
return c;
```

anwar6953 replied:
Updated; thanks for the suggestion.
However, I'm not sure that the performance benefit in this http handler is worth the increased code complexity.

bdmogal commented:
Let's put this on hold. The corresponding PR in CDAP has been reverted in https://github.com/caskdata/cdap/pull/6791

bdmogal commented:
The decision was the make the ClientConfig thread-safe without making it immutable. This PR can be merged after that is fixed.

chtyim commented on src/main/java/co/cask/tracker/AuditTagsHandler.java:
```diff
@@ -53,6 +54,7 @@
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
```
unused import?

anwar6953 commented:
Added these changes (removing unused imports) and squashed commits:
![image](https://cloud.githubusercontent.com/assets/2440977/20450158/197897ac-ada3-11e6-90d6-2abc10345fde.png)

chtyim commented:
LGTM
